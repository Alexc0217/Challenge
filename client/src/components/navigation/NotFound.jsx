import React from "react";
import { Main } from "../ui/styles";
import styled from "styled-components";

const Message = styled.div`
  width: 100%;
  display: flex;
  justify-content: center;
  font-size: 44px;
  font-family: sans-serif;
  padding: 110px 0px;
`

export default function NotFound(){

  return(
    <Main>
      <Message>Página não encontrada :(</Message>
    </Main>
  )
}