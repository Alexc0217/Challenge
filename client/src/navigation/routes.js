const COMPANY = {
  INDEX: "/companies",
  NEW: "/companies/new",
  SHOW: "/companies/:id"
}

const EMPLOYEE = {
  NEW: "/employees/new/:company_id",
  PAIRS: "/employees/pairs/:id",
  SUBORDINATES: "/employees/subordinates/:id",
  SECOND_LEVEL: "/employees/subordinates-second-level/:id"
}

export {
  COMPANY,
  EMPLOYEE
}