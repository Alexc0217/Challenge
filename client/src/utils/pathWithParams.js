export const pathWithParams = (url, params) => {
  let replacedUrl = url;
  
  Object.keys(params).forEach(key => {
    const value = params[key];
    replacedUrl = replacedUrl.replace(`:${key}`, value);
  });

  return replacedUrl;
};
