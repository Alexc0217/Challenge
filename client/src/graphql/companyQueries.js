import { gql } from '@apollo/client';

export const GET_COMPANY = gql`
  query GetCompany($id: ID!){
    company(id: $id){
      name
      id
    }
  }
`

export const CREATE_COMPANY = gql`
  mutation CreateCompany($name: String!) {
    createCompany(input: { name: $name }) {
      company {
        id
        name
      }
      message
      errors
    }
  }
`;

export const COMPANIES = gql`
  {
   companies{
      id
      name
      employeesNumber
    }  
  }
`

export const COMPANY_EMPLOYEES = gql`
  query CompanyEmployee($id: ID!){
    companyEmployees(companyId: $id){
      companyName
      employees{
        name
        id
        role
        email
        managerId
      }
    }
  }
`