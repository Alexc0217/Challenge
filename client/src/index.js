import React from 'react';
import ReactDOM from 'react-dom/client';
import { BrowserRouter as Router, Route, Routes } from 'react-router-dom';
import App from './pages/App';
import { client } from './graphql/client';
import { ApolloProvider } from '@apollo/client';

const root = ReactDOM.createRoot(document.getElementById('root'));
root.render(
  <ApolloProvider client={client}>
    <Router>
      <Routes>
        <Route element={<App />} path="/*" />
      </Routes>
    </Router>
  </ApolloProvider>
);