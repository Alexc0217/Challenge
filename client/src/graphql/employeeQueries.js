import { gql } from '@apollo/client';

export const CREATE_EMPLOYEE = gql`
  mutation CreateEmployee($name: String!, $email: String!, $role: String!, $companyId: ID!){
    createEmployee(input: {name: $name, email: $email, role: $role, companyId: $companyId}){
      employee{
        name
        email
        role
      }
      message
      errors
    }
  }
`

export const DELETE_EMPLOYEE = gql`
  mutation DeleteEmployee($id: ID!){
    deleteEmployee(input: {id: $id}){
      employee{
        name
        id
      }
      message
      errors
    }
  }
`

export const GET_EMPLOYEE_PAIRS = gql`
  query Employee($id: ID!){
    employee(id: $id){
      name
      id
      role
      pairs{
        name
        id
        role
      }
    }
  }
`

export const GET_EMPLOYEE_SUBORDINATES = gql`
  query Employee($id: ID!){
    employee(id: $id){
      name
      id
      role
      subordinates{
        name
        id
        role
      }
    }
  }
`

export const GET_EMPLOYEE_SECOND_LEVEL = gql`
  query Employee($id: ID!){
    employee(id: $id){
      name
      id
      role
      subordinates{
        name
        id
        email
        role
        subordinates{
          id
          name
          email
          role
        }
      }
    }
  }
`

export const UPDATE_EMPLOYEE_MANAGER = gql`
  mutation UpdateEmployeeManager($employeeId: ID!, $managerId: ID!) {
    updateEmployeeManager(input: {employeeId: $employeeId, managerId: $managerId}) {
      employee{
        name
        id
      }
      message
      errors
    }
  }
`;