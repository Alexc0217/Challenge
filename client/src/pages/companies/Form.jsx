import {React, useState} from "react";
import { useMutation } from '@apollo/client';
import {SwalError, SwalSuccess } from "../../components/ui"
import { CREATE_COMPANY } from "../../graphQL/company_queries";
import { Button, CircularProgress } from "@mui/material";
import { Container, DivForm, FormStyled } from "../../components/ui/styles";
import { CustomInput } from "../../components/ui/CustomInput";

function Form() {
  const [name, setName] = useState('');

  const [createCompany, {data, loading, error}] = useMutation(CREATE_COMPANY, {
    onCompleted: (response) => {
      const data = response.createCompany;

      if(data.errors.length > 0) return SwalError(data);
      SwalSuccess({message: data.message, redirect: "/companies"});
    },
    onError: (error) => {
      SwalError(error.message);
    }
  });

  const handleSubmit = (e) => {
    e.preventDefault();
    createCompany({ variables: { name } });
  };

  if (loading) return <CircularProgress />;

  return (
    <DivForm>
      <FormStyled onSubmit={handleSubmit}>
        <CustomInput 
          name={name} 
          onChange={(e) => setName(e.target.value)} 
          label="Nome da empresa" 
          placeholder="Nome da empresa" 
          required 
        />
        <Button type="submit" variant="contained" color="success">Criar empresa</Button>
      </FormStyled>
    </DivForm>  
  );
}

export default Form;
