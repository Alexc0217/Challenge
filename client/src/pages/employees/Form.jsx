import {React, useState} from "react";
import { useMutation } from '@apollo/client';
import {SwalError, SwalSuccess } from "../../components/ui"
import { CREATE_EMPLOYEE } from "../../graphQL/employeeQueries";
import { useParams } from "react-router-dom";
import { Button } from "@mui/material";
import { DivForm, FormStyled } from "../../components/ui/styles";
import { CustomInput } from "../../components/ui/CustomInput";
import Loading from "../../components/ui/Loading";
import { COMPANY } from "../../navigation/routes";
import { pathWithParams } from "../../utils/pathWithParams";

function Form() {
  const [name, setName] = useState('');
  const [email, setEmail] = useState('');
  const [role, setRole] = useState('');
  const [invalidEmail, setInvalidEmail] = useState(false);
  const { company_id } = useParams();

  const [createEmployee, {data, loading, error}] = useMutation(CREATE_EMPLOYEE, {
    onCompleted: (response) => {
      const data = response.createEmployee;
      
      if(data.errors.length > 0) return SwalError(data.errors);
      SwalSuccess({message: data.message, redirect: pathWithParams(COMPANY.SHOW, {id: company_id}) });
    },
    onError: (error) => {
      SwalError(error.message);
    }
  });

  const handleSubmit = (e) => {
    e.preventDefault();
    createEmployee({ variables: { name, email, role, companyId: company_id } });
  };

  if (loading) return <Loading />;

  return (
    <DivForm>
      <FormStyled onSubmit={handleSubmit}>
        <CustomInput 
          name={name} 
          onChange={(e) => setName(e.target.value)} 
          label="Nome" 
          placeholder="Nome" 
          required 
        />
        <CustomInput 
          name={email} 
          onChange={(e) => setEmail(e.target.value)} 
          label="E-mail" 
          placeholder="E-mail" 
          type="email" 
          required
          error={(err) => setInvalidEmail(err)}
        />
        <CustomInput 
          name={role} 
          onChange={(e) => setRole(e.target.value)} 
          label="Cargo" 
          placeholder="Cargo" 
        />
        <Button type="submit" variant="contained" color="success" disabled={invalidEmail} id="submit">Criar Colaborador</Button>
      </FormStyled>
    </DivForm>
  );
}

export default Form;
