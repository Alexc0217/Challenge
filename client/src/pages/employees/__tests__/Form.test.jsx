import React from 'react';
import renderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';
import { MemoryRouter } from 'react-router-dom';
import Form from '../Form';
import { CREATE_EMPLOYEE } from '../../../graphql/employeeQueries';
import { pathWithParams } from '../../../utils/pathWithParams';
import { EMPLOYEE } from '../../../navigation/routes';

const mocks = [
  {
    request: {
      query: CREATE_EMPLOYEE,
      variables: {
        name: 'Alex Moura',
        email: 'alex.moura@outlook.com',
        role: 'Developer',
        companyId: '1'
      },
    },
    result: {
      data: {
        createEmployee: {
          errors: [],
          message: 'Employee created successfully',
        },
      },
    },
  },
];

describe('Form Component Snapshot', () => {
  it('matches the snapshot', () => {
    const tree = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <MemoryRouter initialEntries={[pathWithParams(EMPLOYEE.NEW, {id: 1})]}>
          <Form />
        </MemoryRouter>
      </MockedProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
