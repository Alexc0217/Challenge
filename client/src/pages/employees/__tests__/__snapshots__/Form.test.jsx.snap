// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component Snapshot matches the snapshot 1`] = `
<div
  className="sc-kkmypM jsPyby"
>
  <form
    className="sc-dkjaqt eEMetd"
    onSubmit={[Function]}
  >
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-bz5ng3-MuiInputBase-root-MuiInput-root"
      label="Nome"
      onClick={[Function]}
      required={true}
      variant="outlined"
    >
      <input
        className="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        name=""
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Nome"
        required={true}
        style={
          Object {
            "fontSize": "20px",
            "height": "40px",
          }
        }
        type="text"
      />
    </div>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-bz5ng3-MuiInputBase-root-MuiInput-root"
      label="E-mail"
      onClick={[Function]}
      required={true}
      variant="outlined"
    >
      <input
        className="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        name=""
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="E-mail"
        required={true}
        style={
          Object {
            "fontSize": "20px",
            "height": "40px",
          }
        }
        type="text"
      />
    </div>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary css-bz5ng3-MuiInputBase-root-MuiInput-root"
      label="Cargo"
      onClick={[Function]}
      variant="outlined"
    >
      <input
        className="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        name=""
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Cargo"
        style={
          Object {
            "fontSize": "20px",
            "height": "40px",
          }
        }
        type="text"
      />
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess css-ke5b6m-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      id="submit"
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      Criar Colaborador
    </button>
  </form>
</div>
`;
